api/src/main/java/net/kyori/adventure/text/width/PixelWidthSourceImpl.java

```diff
@@ -0,0 +1,368 @@
+
+  private final Function<CX, CharToIntFunction<Style>> function;
+  private final Function<CX, Locale> localeFunction;
+
+  /**
+   * Get the pixel width of the given character in the Minecraft font, excluding the space between
+   * characters and drop shadow. Handles alphanumerics and most common english punctuation.
+   */
+  static final CharToIntFunction<Style> DEFAULT_FONT_WIDTH = (c, style) -> {
+    int i;
+    if(Character.isUpperCase(c)) {
+      i = c == 'I' ? 3 : 5;
+    } else if(Character.isDigit(c)) {
+      i = 5;
+    } else if(Character.isLowerCase(c)) {
+      switch (c) {
+        case 'i':
+          i = 1;
+          break;
+
+        case 'l':
+          i = 2;
+          break;
+
+        case 't':
+          i = 3;
+          break;
+
+        case 'f':
+        case 'k':
+          i = 4;
+          break;
+
+        default:
+          i = 5;
+          break;
+      }
+    } else {
+      switch (c) {
+        case '!':
+        case '.':
+        case ',':
+        case ';':
+        case ':':
+        case '|':
+          i = 1;
+          break;
+
+        case '\'':
+          i = 2;
+          break;
+
+        case '[':
+        case ']':
+        case ' ':
+          i = 3;
+          break;
+
+        case '*':
+        case '(':
+        case ')':
+        case '{':
+        case '}':
+        case '<':
+        case '>':
+          i = 4;
+          break;
+
+        case '@':
+          i = 6;
+          break;
+
+        default:
+          i = 5;
+          break;
+      }
+    }
+
+    if(style.hasDecoration(TextDecoration.BOLD)){
+      i++;
+    }
+
+    return i;
+  };
+
+  /**
+   * Creates a pixel width source with a function used for getting a {@link CharToIntFunction}.
+   *
+   * <p>Any {@link CharToIntFunction} returned by the function should accept at least all
+   * english alphanumerics and most punctuation and handle {@link TextDecoration#BOLD} in the style.
+   * See {@link PixelWidthSourceImpl#DEFAULT_FONT_WIDTH} for an example.</p>
+   *
+   * @param function a function that can provide a {@link CharToIntFunction} given a context
+   * @since 4.5.0
+   */
+  PixelWidthSourceImpl(final @NonNull Function<CX, CharToIntFunction<Style>> function, final @NonNull Function<CX, Locale> localeFunction){
+    this.function = function;
+    this.localeFunction = localeFunction;
+  }
+
+  @Override
+  public int width(final @NonNull Component component, final @NonNull CX context) {
+    final Map<String, Style> map = this.serialize(component, context);
+    final AtomicInteger length = new AtomicInteger(0);
+
+    map.forEach((string, style) -> string.chars().forEach(c -> length.getAndAdd(this.width((char) c, style, context))));
+
+    return length.get();
+  }
+
+  @Override
+  public int width(final @NonNull String string, final @NonNull Style style, final @NonNull CX context) {
+    int length = 0;
+    for(final char c : string.toCharArray()) length += this.width(c, style, context);
```

Comment:
this won't handle code points outside of the BMP properly

Reply:
(basic multilingual plane)

Reply:
If handling of supplementary characters is needed we can't use `char` in the functions calculating the width for each symbol, which makes it less intuitive for people to make their own (for resource packs++). Do we actually need to support that?

Reply:
People all around the world play Minecraft, and they usually like to have things work correctly when they speak their language.